### PR TITLE
ray exec and ray attach commands

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

`ray exec CLUSTER CMD [--screen] [--start] [--stop]`
`ray attach CLUSTER [--start]`

Example:
`ray exec sgd.yaml 'source activate tensorflow_p27 && cd ~/ray/python/ray/rllib && ./train.py --run=PPO --env=CartPole-v0' --screen --start --stop`

This will in one command create a cluster and run the command on it in a screen session. The screen can later be attached to via ray attach. After the command finishes, the cluster workers will be terminated and the head node stopped.

### CODE REVIEW

**Comment by richardliaw** [comment]
Ran: `ray create_or_update mini.yaml --yes --run 'echo "hello world"'` to start a new cluster, and then got: 

```
NodeUpdater: running

	echo "hello world"

 on None...
ssh: Could not resolve hostname none: nodename nor servname provided, or not known
Traceback (most recent call last):
  File "/Users/rliaw/miniconda3/envs/ray/bin/ray", line 11, in <module>
    load_entry_point('ray', 'console_scripts', 'ray')()
  File "/Users/rliaw/Research/riselab/ray/python/ray/scripts/scripts.py", line 426, in main
    return cli()
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/rliaw/Research/riselab/ray/python/ray/scripts/scripts.py", line 397, in create_or_update
    no_restart, yes, run)
  File "/Users/rliaw/Research/riselab/ray/python/ray/autoscaler/commands.py", line 46, in create_or_update_cluster
    get_or_create_head_node(config, no_restart, yes, run)
  File "/Users/rliaw/Research/riselab/ray/python/ray/autoscaler/commands.py", line 184, in get_or_create_head_node
    updater.ssh_cmd(run, verbose=True, allocate_tty=True)
  File "/Users/rliaw/Research/riselab/ray/python/ray/autoscaler/updater.py", line 207, in ssh_cmd
    stderr=redirect or self.stderr)
  File "/Users/rliaw/miniconda3/envs/ray/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ssh', '-tt', '-o', 'ConnectTimeout=120s', '-o', 'StrictHostKeyChecking=no', '-i', '/Users/rliaw/.ssh/ray-autoscaler_2_us-east-1.pem', 'ubuntu@None', 'bash --login -c \'set -i || true && source ~/.bashrc && echo "hello world"\'']' returned non-zero exit status 255.
```

**Comment by ericl** [comment]
Hm works for me: `ray create_or_update sgd.yaml --yes --run='source activate tensorflow_p27 && cd ~/ray/python/ray/rllib && ./train.py --run=PPO --env=CartPole-v0' --screen`

**Comment by ericl** [comment]
@hartikainen would you like to try this out?

**Comment by hartikainen** [comment]
@ericl definitely! I'll give it a try this week and let you know how it goes. Thanks!

**Comment by hartikainen** [comment]
One thought related to this: It would be awesome to have an option to automatically tear down the cluster once all the experiments are done.

**Comment by ericl** [comment]
@hartikainen added a `--halt` flag that just terminates workers and stops (shutdown -h) the head, see if this works for you.

**Comment by robertnishihara** [comment]
This should all be documented in the API docs http://ray.readthedocs.io/en/latest/api.html#the-ray-command-line-api. The relevant file is https://github.com/ray-project/ray/blob/master/doc/source/api.rst.

File: doc/source/using-ray-on-a-cluster.rst [hunk 1]
```diff
@@ -1,9 +1,9 @@
-Using Ray on a Cluster
-======================
+Manual Cluster Setup
+====================
 
 .. note::
 
-    If you're using AWS you can use the automated `setup commands <http://ray.readthedocs.io/en/latest/autoscaling.html>`__.
+    If you're using AWS or GCE you should use the automated `setup commands <http://ray.readthedocs.io/en/latest/autoscaling.html>`__.
```
Comment (robertnishihara) [inline doc/source/using-ray-on-a-cluster.rst]: I think we should be consistent and use `GCP` throughout (as opposed to `GCE`). What do you think? @hartikainen

**Comment by ericl** [comment]
@hartikainen added a `--cluster-name` flag, open to ideas on how to make this stuff more configurable on the command line.

File: python/ray/autoscaler/commands.py [hunk 2]
```diff
@@ -180,18 +188,76 @@ def get_or_create_head_node(config, no_restart, yes):
...
 
-def get_head_node_ip(config_file):
+def attach_cluster(config_file, start, override_cluster_name):
+    """Attaches to a screen for the specified cluster."""
+
+    exec_cluster(config_file, "screen -L -xRR", False, False, start,
+                 override_cluster_name)
+
+
+def exec_cluster(config_file, cmd, screen, stop, start, override_cluster_name):
```
Comment (robertnishihara) [inline python/ray/autoscaler/commands.py]: The arguments for the functions in this file should be documented in the docstrings.

**Comment by robertnishihara** [comment]
Any reason for choosing `screen` over `tmux`?

File: python/ray/autoscaler/commands.py [hunk 3]
```diff
@@ -180,18 +188,76 @@ def get_or_create_head_node(config, no_restart, yes):
                                        provider.external_ip(head_node)))
 
 
-def get_head_node_ip(config_file):
+def attach_cluster(config_file, start, override_cluster_name):
```
Comment (robertnishihara) [inline python/ray/autoscaler/commands.py]: Any reason for choosing `screen` over `tmux`?

File: doc/source/using-ray-on-a-cluster.rst (hunk 1 above)
Comment (hartikainen) [reply doc/source/using-ray-on-a-cluster.rst]: I think GCP is the same for google as AWS is for amazon, and GCE is the same for google as EC2 is for amazon. So here I would use GCP.

**Comment by richardliaw** [comment]
Can you change the `ssh -i ... 'tail -f /tmp/` to use `ray exec [cluster]`?

**Comment by richardliaw** [comment]
Another wish:  Right now, something like `git status` via `exec` takes very long - would be nice if the boto checks like `Role not specified for head node, using ...` were cached.

**Comment by hartikainen** [comment]
I just tried running my first experiment with this. For some reason the screen attached directly from my local terminal, and it seemed like the command did not get run. Here's are some logs:

```
~/code/softqlearning-private $ ray exec ./config/gcp-ray-autoscaler-mujoco.yaml 'python -m examples.mujoco_all_ray --mode=not-local --env=rllab-swimmer-default --exp_name=20180807-ray-exec-test-1' --cluster-name=20180807-ray-exec-test-1 --screen --start --stop
Cached settings: /var/folders/_l/_msj97l55wsghyt1qh_r0jyh0000gn/T/ray-config-e30e4cb33e501f91d71d7f27aa0676f8c22fa0f2
Launching new head node...
Waiting for operation operation-1533696000959-572e36d54a21a-ce36850a-5c6b3bac to finish...
Done.
Updating files on head node...
NodeUpdater: Updating ray-20180807-ray-exec-test-1-head-4f3bf654 to 249c6048442e42c8854671a93949aa72c50735d3, logging to (console)
Waiting for operation operation-1533696008486-572e36dc77c70-d719e0a9-f8df41ae to finish...
Done.
NodeUpdater: Waiting for IP of ray-20180807-ray-exec-test-1-head-4f3bf654...
NodeUpdater: Waiting for SSH to ray-20180807-ray-exec-test-1-head-4f3bf654...
NodeUpdater: SSH not up, retrying: (Exit Status 255):

	ssh -o ConnectTimeout=5s -o StrictHostKeyChecking=no -i /Users/kristian/.ssh/ray-autoscaler_gcp_us-west1_sac-ray-test_ubuntu.pem ubuntu@35.233.199.98 bash --login -c 'set -i || true && source ~/.bashrc && uptime'


NodeUpdater: Waiting for SSH to ray-20180807-ray-exec-test-1-head-4f3bf654...
NodeUpdater: SSH not up, retrying: (Exit Status 255):

	ssh -o ConnectTimeout=5s -o StrictHostKeyChecking=no -i /Users/kristian/.ssh/ray-autoscaler_gcp_us-west1_sac-ray-test_ubuntu.pem ubuntu@35.233.199.98 bash --login -c 'set -i || true && source ~/.bashrc && uptime'


NodeUpdater: Waiting for SSH to ray-20180807-ray-exec-test-1-head-4f3bf654...
Waiting for operation operation-1533696038203-572e36f8cee7a-f15a82fc-55ef6f53 to finish...
Done.
NodeUpdater: Syncing /Users/kristian/code/softqlearning-private/.git/refs/heads/master to /tmp/current_branch_sha...
building file list ... done
master

sent 167 bytes  received 42 bytes  139.33 bytes/sec
total size is 41  speedup is 0.20
NodeUpdater: Syncing /Users/kristian/code/softqlearning-private/.mujoco/mjkey.txt to ~/rllab/vendor/mujoco/mjkey.txt...
building file list ... done

sent 87 bytes  received 20 bytes  214.00 bytes/sec
total size is 965  speedup is 9.02
NodeUpdater: Syncing /Users/kristian/.ssh/ray-autoscaler_gcp_us-west1_sac-ray-test_ubuntu.pem to ~/ray_bootstrap_key.pem...
building file list ... done

sent 129 bytes  received 20 bytes  99.33 bytes/sec
total size is 1679  speedup is 11.27
NodeUpdater: Syncing /var/folders/_l/_msj97l55wsghyt1qh_r0jyh0000gn/T/ray-bootstrap-0zpct1ti to ~/ray_bootstrap_config.yaml...
building file list ... done
ray-bootstrap-0zpct1ti

sent 714 bytes  received 66 bytes  520.00 bytes/sec
total size is 2157  speedup is 2.77
Waiting for operation operation-1533696049567-572e3703a5518-8f405492-51c592f3 to finish...
NodeUpdater: running

	(cd ~/softqlearning-private && git fetch && git checkout `cat /tmp/current_branch_sha`

 on 35.233.199.98...
Note: checking out 'aead5833820fde531293e637816f434d2db1a724'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at aead583... Add log_dir print in launch_experiment function
NodeUpdater: running

	ray stop

 on 35.233.199.98...


global_scheduler: no process found
plasma_store: no process found
plasma_manager: no process found
local_scheduler: no process found
raylet: no process found
raylet_monitor: no process found
NodeUpdater: running

	ray start --head --redis-port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml

 on 35.233.199.98...
Process STDOUT and STDERR is being redirected to /tmp/raylogs/.
Waiting for redis server at 127.0.0.1:6379 to respond...
Waiting for redis server at 127.0.0.1:21599 to respond...
Starting local scheduler with the following resources: {'CPU': 16, 'GPU': 0}.
Using IP address 10.138.0.3 for this node.
{'object_manager_ports': [8076], 'node_ip_address': '10.138.0.3', 'redis_address': '10.138.0.3:6379', 'object_store_addresses': [ObjectStoreAddress(name='/tmp/plasma_store86301715', manager_name='/tmp/plasma_manager54920249', manager_port=8076)], 'local_scheduler_socket_names': ['/tmp/scheduler78450238'], 'raylet_socket_names': [], 'webui_url': 'http://localhost:8888/notebooks/ray_ui26905.ipynb?token=254998b82adf7a8b3eccc2f2a8e806c533bce97dd6eb538e'}

Started Ray on this node. You can add additional nodes to the cluster by calling

    ray start --redis-address 10.138.0.3:6379

from the node you wish to add. You can connect a driver to the cluster from Python by running

    import ray
    ray.init(redis_address="10.138.0.3:6379")

If you have trouble connecting from a different machine, check that your firewall is configured properly. If you wish to terminate the processes that have been started, run

    ray stop

======================================================================
View the web UI at http://localhost:8888/notebooks/ray_ui26905.ipynb?token=254998b82adf7a8b3eccc2f2a8e806c533bce97dd6eb538e
======================================================================

Waiting for operation operation-1533696075693-572e371c8fbc8-9d40438b-dd3bd7c3 to finish...
Done.
NodeUpdater: Applied config 249c6048442e42c8854671a93949aa72c50735d3 to node ray-20180807-ray-exec-test-1-head-4f3bf654
Head node up-to-date, IP address is: 35.233.199.98
To monitor auto-scaling activity, you can run:

  ray exec ./config/gcp-ray-autoscaler-mujoco.yaml 'tail -n 100 -f /tmp/raylogs/monitor-*'

To login to the cluster, run:

  ssh -i /Users/kristian/.ssh/ray-autoscaler_gcp_us-west1_sac-ray-test_ubuntu.pem ubuntu@35.233.199.98

NodeUpdater: running

	screen -L -dm bash -c 'python -m examples.mujoco_all_ray --mode=not-local --env=rllab-swimmer-default --exp_name=20180807-ray-exec-test-1; ray stop; ray teardown ~/ray_bootstrap_config.yaml --yes --workers-only; sudo shutdown -h now; exec bash'
  
 on 35.233.199.98...
```

Here, the screen attached to my terminal. I was expecting the ray exec command to just return and have to manually attach to the head. It also didn't seem like the command got run.

```
ubuntu@ray-20180807-ray-exec-test-1-head-4f3bf654:~$
ubuntu@ray-20180807-ray-exec-test-1-head-4f3bf654:~$ exit
```

Finally some logs after detaching the screen.
```
Connection to 35.233.199.98 closed.
There is a screen on:
	4311..ray-20180807-ray-exec-test-1-head-4f3bf654	(08/08/2018 02:40:09 AM)	(Dead ???)
Remove dead screens with 'screen -wipe'.
[screen is terminating]
Connection to 35.233.199.98 closed.
```

**Comment by ericl** [comment]
I've also seen the same issue. Oddly enough, if you retry it a few times
then it starts working. I was hoping it was a problem specific to my
machine configuration but perhaps not.

Maybe there's something about freshly started nodes that is causing screen
to crash (the command doesn't appear to get run at all and there is no
logging output).

File: python/ray/autoscaler/commands.py (hunk 2 above)
Comment (ericl) [reply python/ray/autoscaler/commands.py]: Done

File: python/ray/autoscaler/commands.py (hunk 3 above)
Comment (ericl) [reply python/ray/autoscaler/commands.py]: I think tmux isn't installed on machines by default.

**Review by ericl — COMMENTED** [review]
This is ready for review

**Comment by ericl** [comment]
@hartikainen I'm going to mark the screen flag as experimental for now. Will try to dig up the root cause later, it's quite an odd bug.

File: python/ray/scripts/scripts.py [hunk 4]
```diff
@@ -380,39 +387,108 @@ def stop():
...
+    click.echo(get_head_node_ip(cluster_config_file, cluster_name))
 
 
 cli.add_command(start)
 cli.add_command(stop)
 cli.add_command(create_or_update)
+cli.add_command(create_or_update, name="up")
+cli.add_command(attach)
+cli.add_command(exec_cmd, name="exec")
 cli.add_command(teardown)
```
Comment (hartikainen) [inline python/ray/scripts/scripts.py]: Should this have `name="down"` to be "constistent" with `name="up"` in the `create_or_update` command?

**Review by hartikainen — APPROVED** [review]
Looks good to me!

File: python/ray/scripts/scripts.py (hunk 4 above)
Comment (ericl) [reply python/ray/scripts/scripts.py]: Good idea, added that as an alias.

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
Hey, to be clear (and public), the _correct_ behavior for the screen flag in `exec` should not attach to the terminal right?

**Comment by ericl** [comment]
It should attach, so that you see it starting. However since it's a screen
you can always just close it whenever.

**Comment by richardliaw** [comment]
OK I guess that's reasonable. Worst case, the user can just run it themselves.

Since screen doesn't work (and literally nothing online provides any insight), I'm currently manually appending (`tmux new -d -s mysession`) to the beginning of my exec command.

Perhaps we should just consider using tmux, and maybe have a flag to auto-install.

**Comment by hartikainen** [comment]
I've been running some experiments with this and overall I'm pretty satisfied! The occasional bug with the screen is a bit annoying so I've mostly ended up running things without screen. That's not a problem though.

One feedback I have is about the automatic detaching on exec. I have actually not yet seen any use for it so far, and I feel like I should at least have an option to run the exec such that the local process just finishes and leaves the command running on the cluster. My preference would be to have this be the default behavior but don't feel super strongly. Any thoughts?

**Comment by ericl** [comment]
I removed the post-screen attach (I suppose you can always ray attach right after, and now that we cache the resolved cluster configs that should be fairly fast).

@richardliaw  I think we can do tmux if it's present, though I'm not familiar with it.

**Comment by hartikainen** [comment]
Is there a way to run this "asynchronously" such that the (local) ray exec python process just exits and the actual cmd keeps running on the cluster? Right now, when I kill the terminal that I ran ray exec from, the cluster head seems to shut down. Is this expected?

**Comment by richardliaw** [comment]
Yeah - @hartikainen what are you running and what is the output on the terminal?

To echo Eric's point, maybe consider running your exec job with TMUX detachment instead of screen for now.

**Comment by hartikainen** [comment]
Yeah, I think what I was looking for was some other "hack" to get the behavior of screen, but I guess it's just easier for me to install tmux on our images. The screen bug is super weird.

**Pull request merged**